Step 3 : 구간 제거 기능 심화 구현

hwihwi523 opened the pull request:
안녕하세요, 리뷰어님!

이번에 Step 3 리뷰 요청 드리게 된 윤재휘입니다!

이번에 구현한 주 내용은 구간 제거 기능과 예외처리입니다. 구현 내용은 다음과 같습니다.

- 구간 제거 기능 심화 : 종점역 삭제는 구간 삭제만 진행하고, 중간역 삭제는 새로운 구간을 추가하는 것까지 해야 합니다. 따라서 아래와 같이 단계를 나누었습니다.
    1. (공통) 삭제할 역이 주어지면 해당 역이 포함된 모든 Section을 검색합니다.
    2. (공통) 1번에서 검색한 Section을 DB에서 모두 삭제합니다.
    3. (중간역) 만약 위에서 삭제한 Section의 개수가 2개라면, 해당 역은 중간에 위치해 있다는 것을 의미합니다. 따라서 삭제할 역의 양옆 역을 이은 새로운 구간을 추가합니다.
- 예외처리
    1. 발생할 수 있는 예외별로 오류코드와 안내 메시지를 만들어 enum에 정의했습니다.
    2. 오류코드를 저장하는 커스텀 예외를 만들어 오류가 발생했을 경우 코드와 메시지를 제공합니다.
    3. 커스텀 예외를 핸들링하는 GlobalExceptionHandler를 만들어, 예외 발생 시 코드와 메시지를 담는 ErrorResponse를 반환하도록 했습니다.

이번 미션을 진행하면서 느낀 질문이 있는데 혹시 답변해 주신다면 정말 감사드리겠습니다!
1. 데이터를 삽입할 때, DB를 직접 조회해서 중복을 검사하는 것과, 일단 삽입하고 DB로부터 예외를 받아서 처리하는 것 중 어느 방식이 더 효율적이고 더 많이 쓰일까요? 제 생각으로는 후자의 방식이 쿼리를 덜 날리기에 더 효율적일 것으로 생각하지만, 예외를 던짐으로써 DB에 부하가 더 생기진 않을까 궁금해서 여쭈어 봅니다!
2. 데이터를 삭제할 때, 삭제하고자 하는 데이터가 DB에 없어도 오류가 발생하지는 않는 것으로 알고 있습니다. 이런 경우 삭제 전에 데이터 존재 여부를 검사하는 게 좋을까요, 아니면 어차피 오류가 나지 않으니 그냥 두는 게 좋을까요?
3. GlobalExceptionHandler 클래스는 패키지 구조 상 어느 곳에 넣는 게 가장 어울릴까요?

시간 내어 읽어주셔서 감사합니다!

hwihwi523 committed:
style: difference() 메서드명을 subtract()로 변경

hwihwi523 committed:
refactor: equals(), hashCode()에서 식별자만 비교하도록 수정

hwihwi523 committed:
test: Distance 간 차이를 구하거나 대소 관계를 비교하는 테스트 코드 작성

hwihwi523 committed:
refactor: Station 비교 시 upStation을 직접 꺼내서 비교하지 않고, Section 내부에서 비교하고 메시지를 전달받도록 수정

hwihwi523 committed:
refactor: upStations, downStations를 필드에서 제외하고, Set으로 변환하는 로직을 메서드로 분리

hwihwi523 committed:
feat: 구간을 새로 등록했을 때 Location 헤더에 새로 등록된 구간의 ID도 추가

hwihwi523 committed:
refactor: 불필요 import 제거 및 불필요 Transactional 어노테이션 삭제

hwihwi523 committed:
docs: 3단계 요구사항 추가

hwihwi523 committed:
test: 존재하는 구간을 삭제하는 경우, 존재하지 않는 구간을 삭제하는 경우에 대한 테스트 작성

hwihwi523 committed:
feat: 존재하는 구간을 삭제하는 경우, 존재하지 않는 구간을 삭제하는 경우에 대한 프로덕션 코드 작성

hwihwi523 committed:
style: 메서드명 수정

hwihwi523 committed:
test: 종점역 제거, 중간 제거 등 구간 제거 테스트 작성

hwihwi523 committed:
feat: 종점역 제거, 중간 제거 등 구간 제거 프로덕션 코드 작성

hwihwi523 committed:
refactor: CreateHelper의 역과 노선을 생성하는 로직 리팩토링

- getStationIds()로 모든 지하철역 ID를 받아 와 구간을 생성하는 부분을 삭제하고, 지하철역을 생성할 때 생성된 ID를 반환받아 구간을 생성하도록 변경

hwihwi523 committed:
refactor: NamedParameterJdbcTemplate으로 변경 및 구간 삭제 시 IN clause 사용

hwihwi523 committed:
style: 변경된 요구사항에 맞게 에러 메시지 수정

hwihwi523 committed:
feat: 오류코드 enum 생성 및 오류코드에 대한 README 내용 작성

hwihwi523 committed:
feat: 커스텀 예외 생성 및 기존 예외처리 코드, 테스트 수정

hwihwi523 committed:
feat: 예외처리 핸들러 생성 및 기존 코드, 테스트 수정

hwihwi523 committed:
feat: 데이터 조회 시 Optional로 반환하도록 적용

hwihwi523 committed:
feat: 지하철 역 삭제 시 데이터 존재 여부를 확인하는 테스트 및 코드 작성

hwihwi523 committed:
feat: Line 조회 시 Optional 적용 및 오류코드 추가

hwihwi523 committed:
test: 존재하지 않는 Line 조회 시 예외를 던지는 테스트 작성

hwihwi523 committed:
test: 지하철 역명, 노선명 등 이름 생성과 검증에 대한 테스트 작성

hwihwi523 committed:
feat: 지하철 역명, 노선명을 저장할 일급 컬렉션 Name 생성 및 적용

hwihwi523 committed:
feat: LineStationsResponse 필드 중 domain.Station 리스트를 dto.StationResponse 리스트로 변경

hwihwi523 committed:
test: 빈 문자열, 최대 길이를 초과한 문자열로 지하철 역과 노선을 생성하는 통합 테스트 작성

hwihwi523 committed:
feat: 빈 문자열, 최대 길이를 초과한 문자열로 지하철 역과 노선을 생성하는 예외처리 구현

hwihwi523 committed:
test: 비어 있거나 최대 길이를 초과한 색상명으로 노선을 생성하는 통합 테스트 작성

hwihwi523 committed:
feat: 비어 있거나 최대 길이를 초과한 색상명으로 노선을 생성하는 예외처리 구현

begaonnuri commented on src/main/java/subway/application/LineService.java:
```diff
@@ -56,14 +68,20 @@ public List<Line> findLines() {
     public Line findLineById(final Long id) {
-        return lineDao.findById(id);
+        return lineDao.findById(id)
+                .orElseThrow(() -> new LineException(ErrorCode.NO_SUCH_LINE, NO_LINE_EXCEPTION_MESSAGE));
```
`ErrorCode.NO_SUCH_LINE`에 정의된 메시지와 `NO_LINE_EXCEPTION_MESSAGE`가 중복되는것같은데 변경될 경우 2개를 신경써야해서 관리가 어렵지 않을까요?

hwihwi523 replied:
우선 두 가지의 메시지를 넣은 이유는 사용자에게 보여줄 메시지와 개발자에게 보여줄 메시지를 분리하기 위함이었습니다!

사실 리뷰어님 말씀대로 두 메시지가 중복되는 예외가 몇 개 있습니다. 그래서 처음에는 중복된 메시지를 갖는 예외는 ErrorCode만 넘기고, 예외 안에서 ErrorCode의 메시지를 그대로 넣는 생각을 했습니다. 그런데 예외를 던지는 다른 로직을 구현할 때, 개발자 메시지만 작성하면 되는데 사용자 메시지를 찾아 중복인지 확인하는 작업이 추가로 필요할 것 같아서 중복을 감수하고 분리했습니다!

begaonnuri commented on src/main/java/subway/application/LineService.java:
```diff
@@ -89,23 +108,58 @@ public void saveSection(final SectionRequest request, final Long lineId) {
 
-        sections.delete(delete);
-        sectionDao.deleteByStation(delete, lineId);
+        if (isMidRemoved(removeSections)) {
+            sectionDao.insert(newSectionAfterRemoval(removeSections), lineId);
+        }
```
데이터베이스에 삽입하는 부분을 제외하면 도메인에서 진행할 수 있지 않을까요? 도메인의 역할이 Service 레이어에서 이루어지는것 같아요.

hwihwi523 replied:
DeleteSections 클래스를 만들고, 중간 구간 삭제인지 확인하는 로직을 추가해 역할을 분리했습니다!

begaonnuri commented on src/main/java/subway/application/LineService.java:
```diff
@@ -89,23 +108,58 @@ public void saveSection(final SectionRequest request, final Long lineId) {
     }
+
+    private Section newSectionAfterRemoval(final List<Section> removeSections) {
+        final Section firstSection = removeSections.get(0);
+        final Section secondSection = removeSections.get(1);
```
removeSections이 반드시 원하는 형태여야지만 동작하는 구조로 보여요. 차라리 2개의 Section을 인자로 받거나 스트림을 활용해 요소를 찾는 방식으로 처리하면 어떨까요?

begaonnuri commented on src/main/java/subway/config/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,65 @@
+package subway.config;
```
PR 내용으로 질문 주신 부분이기도 한데, MVC 형태가 아닌 아키텍처도 많아서 확답을 드릴 순 없지만 일반적으론 controller와 유사한 위치에 두거나 exception을 관리하는 패키지를 두기도 해요.

패키지 구조에 정답은 없습니다. 하지만 config는 `@Configuration`을 통한 설정 클래스가 올 것 같아 변경하는게 좋아 보여요.

begaonnuri commented on src/main/java/subway/config/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,65 @@
+
+        ErrorCode errorCode;
+
+        if (e instanceof LineException) {
+            errorCode = ((LineException) e).getErrorCode();
+        } else if (e instanceof SectionException) {
+            errorCode = ((SectionException) e).getErrorCode();
+        } else if (e instanceof StationException) {
+            errorCode = ((StationException) e).getErrorCode();
+        } else {
+            errorCode = ((InvalidRequestException) e).getErrorCode();
+        }
```
에러가 늘어날때마다 선형적으로 코드가 늘어날 수 있지 않을까요?

hwihwi523 replied:
기존 방식에서는 RuntimeException을 상속받다보니 ErrorCode로 접근하기 위해 위와 같이 구성해야 했습니다.

그래서 ErrorCode를 갖는 CustomException을 만들고, Section, Station 등 각 예외가 CustomException을 상속받도록 했습니다.

이후 핸들러의 인자 타입을 CustomException으로 설정해서 바로 ErrorCode에 접근할 수 있도록 다형성을 활용했습니다!

begaonnuri commented on src/main/java/subway/config/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,65 @@
+            final DataIntegrityViolationException e) {
+        log.error(e.getMessage());
+
+        return ResponseEntity.badRequest().body(ErrorResponse.of(ErrorCode.EXISTS_DATA));
```
데이터베이스 연산에 실패한 로직이 4xx 에러라고 생각하신 이유가 있을까요?

hwihwi523 replied:
이미 존재하는 노선명이나 역명 등 중복 키가 있을 때 발생하는 예외기 때문에, 사용자 요청이 잘못된 것이라 판단했습니다!

begaonnuri commented on src/main/java/subway/dao/SectionDao.java:
```diff
@@ -1,63 +1,99 @@
-    public void deleteByStation(final Station station, final Long lineId) {
-        String sql = "delete from SECTION where down_station_id = ? and line_id = ?";
+    public void deleteAllIn(final List<Long> ids) {
+        final String sql = "delete from SECTION where id in (:removeIds)";
```
in 절을 잘 활용해주셨네요 👍🏻

begaonnuri commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -1,29 +1,36 @@
     private final int value;
 
-    public Distance(final int value) {
+    public Distance(final Integer value) {
```
래퍼클래스를 사용하신 이유가 있을까요?

hwihwi523 replied:
거리값으로 null이 들어올 가능성을 고려했습니다!

그런데 null 입력에 대한 처리는 도메인보다 앞단에서 먼저 처리하는 게 좋을 것 같아서, request 검증 로직을 추가하고 다시 primitive 타입으로 변경했습니다!

begaonnuri commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -1,50 +1,52 @@
 
 public class Sections {
 
-    private static final String SAME_STATION_EXCEPTION_MESSAGE = "새로운 구간의 상행역은 기존 하행 종점역과 같아야 합니다.";
+    private static final String INVALID_STATIONS_EXCEPTION_MESSAGE = "삽입 시 기준역은 한 개의 역이어야 합니다.";
     private static final String EMPTY_EXCEPTION_MESSAGE = "최소 1개 이상의 구간이 있어야 합니다.";
     private static final String AT_LEAST_ONE_SECTION_EXCEPTION_MESSAGE = "구간은 0개가 될 수 없습니다.";
-    private static final String DELETE_ONLY_LAST_SECTION_EXCEPTION_MESSAGE = "마지막 구간만 삭제할 수 있습니다.";
     private static final String CANNOT_FIND_START_SECTION_EXCEPTION_MESSAGE = "출발역에 해당되는 구간을 찾을 수 없습니다.";
     private static final String TWO_MORE_START_STATION_EXCEPTION_MESSAGE = "상행 종점역이 두 개 이상입니다.";
     private static final String CANNOT_FIND_START_STATION_EXCEPTION_MESSAGE = "상행 종점역을 찾을 수 없습니다.";
     private static final String LONGER_THAN_OLDER_SECTION_EXCEPTION_MESSAGE = "삽입하는 새로운 구간의 거리는 기존 구간보다 짧아야 합니다.";
+    private static final String NOT_EXIST_STATION_EXCEPTION_MESSAGE = "삭제할 역이 존재하지 않습니다.";
+    private static final String CANNOT_FIND_OLD_SECTION = "기존 구간을 찾을 수 없습니다.";
```
> 상수는 대체로 어떤 상황에서도 변하지 않는 값들을 사용합니다.
저는 예외 메시지도 이에 해당한다고 생각했고, 메시지들을 상수로 선언하고 모아서 바로 확인할 수 있다면 관리도 더욱 용이할 거라 생각했습니다!

어떤 점이 용이하다고 느끼셨나요? 모아서 확인하면 어떤 점이 좋나요?

hwihwi523 replied:
어떤 예외를 처리했는지 바로 확인할 수 있고, 해당 예외를 던지는 코드도 빠르게 찾을 수 있다는 점에서 용이하다고 느꼈습니다!

혹시 현업에서 이러한 예외 메시지는 상수로 선언하지 않고 그대로 쓰는 편인가요???

begaonnuri replied:
네 현업에선 특정 경우(국제화 등)을 제외하면 저렇게 사용하진 않는 편이에요. 뿐만 아니라 다양한 라이브러리의 코드도 한번 확인해보시면 좋을 것 같아요.

클래스를 확인할때 예외를 확인하는 경우 보다는 클래스가 담고 있는 필드나 메서드를 확인하는 경우가 대부분인데 저렇게 상수로 많은 부분을 차지하고있으면 컨텍스트를 파악하기 쉽지 않다고 생각해요.

그리고 상수라는건 이 값이 어떤 것을 의미하는지 표현하거나 관련된 값을 한번에 관리하기 쉽게 하기 위해 사용한다고 생각하는데 지금과 같은 메시지는 재사용하지도 않을 뿐더러 상수 명은 단순히 메시지를 간략하게 영어로 옮긴것에 불과하지 않아 필요 없다고 생각해요.  재휘님 생각은 어떠신가요?

hwihwi523 replied:
재사용하지 않는 메시지, 간략하게 영어로 옮긴 것뿐이라는 말씀이 굉장히 와 닿습니다!

말씀을 듣고 다시 생각해 보고 코드도 다시 보니 단점이 많이 보였습니다.

우선 상수 메시지가 많은 클래스에서는 상수 메시지가 필드의 대부분을 차지했고, MIN_xxx, MAX_xxx와 같이 의미 있는 상수를 파악하기 힘들었습니다.
그리고 단순히 영어로 옮긴 것이기 때문에 '영어 -> 한글 -> 의미 파악'처럼 한 단계를 더 거쳐야 했습니다.

현재는 상수를 삭제하고 모두 예외에서 그대로 문자열을 받도록 수정했습니다. 감사합니다!!

begaonnuri commented on src/main/java/subway/domain/Station.java:
```diff
@@ -5,25 +5,30 @@
 
     private Long id;
-    private String name;
+    private Name name;
```
노선 이름과 역 이름의 정책이 달라지면 어떻게 될까요?

hwihwi523 replied:
생각해보니 정책이 달라지면 Name 클래스 하나로 노선명과 역명 모두 적용할 수 없게 되네요!

그래서 StationName, LineName으로 분리하여 개별적으로 적용할 수 있도록 수정했습니다!

begaonnuri commented on src/main/java/subway/exception/ErrorCode.java:
```diff
@@ -0,0 +1,60 @@
+package subway.exception;
+
+public enum ErrorCode {
```
에러 코드는 어떤 목적으로 사용해주신걸까요?

hwihwi523 replied:
서버와 클라이언트 간 오류코드에 대한 규약을 둬서 오류를 더욱 간편히 찾고자 했습니다.

만약 클라이언트에서 오류가 발생해 문의가 들어왔을 때, 오류코드를 전달받는다면 어느 오류인지 금방 파악할 수 있을 것 같았습니다!

begaonnuri commented on src/main/java/subway/exception/ErrorCode.java:
```diff
@@ -0,0 +1,60 @@
+package subway.exception;
+
+public enum ErrorCode {
+    INVALID_COLOR_LENGTH(400, "COL001", "색상명의 길이를 벗어났습니다."),
```
COL001와 같은 코드의 용도는 무엇인가요?

hwihwi523 replied:
위와 같이 오류코드를 만들어 오류를 쉽게 찾고자 했습니다.

코드 이름은 가상으로 만들었는데 기능명과 숫자의 조합으로 구성했습니다.

begaonnuri commented on src/main/java/subway/exception/InvalidRequestException.java:
```diff
@@ -0,0 +1,15 @@
+package subway.exception;
+
+public class InvalidRequestException extends RuntimeException {
```
커스텀 예외를 사용하시면서 어떤 점이 좋다고 느껴지셨나요?

hwihwi523 replied:
커스텀 예외를 어떤 기준으로 만드는지 아직 정확히는 잘 모르겠습니다..!

그러나 이번 미션에서는 Section, Line 등 기능으로 커스텀 예외를 만든다면, 예외가 발생했을 때 어느 기능에 문제가 있는지 더욱 잘 확인할 수 있을 것 같았습니다!

begaonnuri replied:
해당 예외 타입을 통해 무언가 처리가 필요한 경우 사용하면 좋다고 생각해요.

지금도 나쁘진 않지만, 개인적으로 지금은 그런 로직이 없으니 단순히 `IllegalXXException`을 사용해도 충분하다고 생각해요. 예외 타입도 결국 관리해야할 클래스라서 많아지면 피로해지실거에요. 참고만 하시고 꼭 변경하진 않으셔도 됩니다!

begaonnuri commented on src/main/java/subway/ui/LineController.java:
```diff
@@ -63,20 +63,19 @@ public ResponseEntity<Void> deleteLine(@PathVariable Long id) {
     @DeleteMapping("/{id}/sections")
     public ResponseEntity<Void> deleteSection(@PathVariable Long id, @RequestParam String stationId) {
-        lineService.deleteSectionByStationId(id, stationId);
+        if (!stationId.strip().matches("\\d+")) {
+            throw new IllegalArgumentException(NOT_POSITIVE_STATION_ID_EXCEPTION_MESSAGE);
+        }
```
프레임워크가 해주는 일이 중복으로 처리되고있네요. stationId의 메서드 시그니처를 Long 타입으로 선언하면 바로 Long 타입으로 받을 수 있습니다 🙂

hwihwi523 replied:
stationId로 숫자가 아닌 문자가 들어올 가능성이 있다고 판단했습니다!

그래서 일단 문자열로 받은 다음, 숫자만 존재하는지 확인하고 Long으로 변환해 주었습니다!

begaonnuri commented on src/test/java/subway/integration/LineIntegrationTest.java:
```diff
@@ -79,6 +79,82 @@ void createLineWithDuplicateName() {
+        final LineRequest emptyLineRequest = new LineRequest("", "bg-red-600", 1L, 2L, 10);
+
+        // when
+        ExtractableResponse<Response> response = RestAssured
+                .given().log().all()
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .body(emptyLineRequest)
+                .when().post("/lines")
+                .then().log().all().
+                extract();
```
`RestAssured`의 스펙으로 인해 내용을 한눈에 파악하기 어려운것 같아요. 이 부분을 따로 분리해보면 어떨까요?

begaonnuri commented on src/test/java/subway/integration/LineIntegrationTest.java:
```diff
@@ -13,28 +18,23 @@
-        CreateHelper.createStation("역삼역");
-        CreateHelper.createStation("교대역");
-        CreateHelper.createStation("강변역");
+        final Long firstStationId = CreateHelper.createStation("잠실역");
+        final Long secondStationId = CreateHelper.createStation("역삼역");
+        final Long thirdStationId = CreateHelper.createStation("교대역");
+        final Long fourthStationId = CreateHelper.createStation("강변역");
```
호불호가 갈리긴 하지만 이런 경우 `잠실역ID`처럼 한글 변수명을 사용하면 더욱 가독성이 올라갑니다. 이 부분은 선택적으로 반영해주세요!

begaonnuri reviewed (changes requested):
안녕하세요 재휘님, 3단계도 잘 구현해주셨네요. 👍🏻 
몇가지 코멘트 남겼는데 확인 부탁드려요.
잘 이해가 안되거나 어렵게 느껴지시면 편하게 DM 주세요 🙂

begaonnuri commented:
https://github.com/next-step/spring-subway/pull/42#issue-1821927226 에 대한 답변입니다.

Q. 데이터를 삽입할 때, DB를 직접 조회해서 중복을 검사하는 것과, 일단 삽입하고 DB로부터 예외를 받아서 처리하는 것 중 어느 방식이 더 효율적이고 더 많이 쓰일까요? 제 생각으로는 후자의 방식이 쿼리를 덜 날리기에 더 효율적일 것으로 생각하지만, 예외를 던짐으로써 DB에 부하가 더 생기진 않을까 궁금해서 여쭈어 봅니다!

A. 프로그램의 로직을 진행시킬때 예외를 발생시키고 catch 할수도 있지만 if 문을 사용하듯, 예외는 로직상 사이드 이펙트를 처리할때 사용하는것이 좋다고 생각해요. 현업에서 복잡한 쿼리가 병목이지 커넥션 한두번을 맺는 것이 병목은 아니기 때문에 크게 신경 쓰지 않는 편이에요. 코드의 가독성을 더 중시하는게 좋다고 생각해요.

Q. 데이터를 삭제할 때, 삭제하고자 하는 데이터가 DB에 없어도 오류가 발생하지는 않는 것으로 알고 있습니다. 이런 경우 삭제 전에 데이터 존재 여부를 검사하는 게 좋을까요, 아니면 어차피 오류가 나지 않으니 그냥 두는 게 좋을까요?

A. 반대로 재휘님이 어떤 사이트에서 회원탈퇴를 했는데 성공적으로 탈퇴되었다고 안내가 나오지만 계정이 여전히 남아있다면 어떤 느낌이 드실 것 같나요? spring-data-jdbc, spring-data-jpa와 같은 오픈소스에서 어떻게 해결했는지를 찾아보는것도 좋은 공부 방법이겠네요!

Q. GlobalExceptionHandler 클래스는 패키지 구조 상 어느 곳에 넣는 게 가장 어울릴까요?

A. 이 부분은 코멘트로 남겼으니 참고해주세요.

hwihwi523 committed:
test: Name을 LineName, StationName으로 분리하고 각각의 테스트 작성

hwihwi523 committed:
refactor: Name을 LineName, StationName으로 분리

hwihwi523 committed:
feat: Request 검증 로직 추가

hwihwi523 committed:
refactor: GlobalExceptionHandler에서 예외마다 캐스팅하는 로직 수정

- RuntimeException에는 ErrorCode에 관한 코드가 없어서 각 예외로 직접 캐스팅해야 했으나, ErrorCode를 갖는 CustomException을 만들고 해당 예외를 상속시켜 다형성으로 캐스팅 대체

hwihwi523 committed:
refactor: 요청을 날리는 RestAssuredHelper 클래스를 만들고, 기존 코드에서 요청 로직을 분리

hwihwi523 committed:
feat: 노선 및 지하철 역을 삽입하거나 삭제하기 전 데이터 존재 여부 검사 로직 추가

hwihwi523 committed:
feat: DeleteSections 클래스를 만들어 기존 서비스에서 구간 삭제 관련 역할 분리 / 구간 삽입 관련 역할 분리

hwihwi523 committed:
refactor: DeleteSections 클래스가 Section 리스트 대신 Sections를 갖도록 수정

hwihwi523 committed:
refactor: 미사용 메서드 삭제 및 개행 개선

hwihwi523 committed:
refactor: RestAssured 로직을 기존에 분리해 둔 메서드로 변경하여 가독성 개선

hwihwi523 commented:
안녕하세요, 리뷰어님!

리뷰해 주신 사항을 토대로 리팩토링 했으며, 그 외 변경사항은 다음과 같습니다!

1. DeleteSections 클래스를 만들어 삭제할 구간을 관리하도록 역할 위임
2. 삽입 전 데이터를 조회해서 중복 데이터 존재 여부 검사
3. 삭제 전 데이터를 조회해서 삭제할 데이터 존재 여부 검사
4. Request 및 Request 필드에 대한 Null 체크 등 유효성 검증 로직 추가

시간 내어 읽어주셔서 감사합니다!

begaonnuri commented on src/main/java/subway/application/LineService.java:
```diff
@@ -117,30 +128,24 @@ public Long saveSection(final SectionRequest request, final Long lineId) {
-        if (isMidRemoved(removeSections)) {
-            sectionDao.insert(newSectionAfterRemoval(removeSections), lineId);
+        if (deleteSections.isKindOfMidDeletion()) {
+            sectionDao.insert(deleteSections.newSection(), lineId);
```
`DeleteSections` 모델을 만들어주셨네요. 의도하신 대로 모델이 사용되려면 `isKindOfMidDeletion` 메서드가 아니라 생성자를 통해 검증이 이루어져야 하지 않을까요? 지금처럼 사용하는 경우엔 `Sections`을 사용하는것과 큰 차이가 없어보여요.

hwihwi523 replied:
DeleteSections는 삭제할 구간 보관, 삭제 유형 반환, 새로 생성되는 구간을 반환하도록 역할을 부여했습니다!

따라서 DeleteSections는 중간 구간을 삭제하는 경우뿐만 아니라 종점을 삭제하기도 해서 생성자로 빼지 않고 위와 같이 메서드로 만들어 사용했습니다!

begaonnuri reviewed (approved):
안녕하세요 재휘님, 리뷰 잘 반영해주셨네요!
이번 단계는 여기서 머지하겠습니다. 다음 단계 진행해주세요 💪🏻
코멘트도 남겼는데 참고해주세요.
